include/nccl_ofi.h

```diff
@@ -247,12 +276,22 @@ struct nccl_net_ofi_device {
 	/* this device's index in the plugin's devices array */
 	int dev_id;
 
-	/* name of the device - should include the provider name, but
-	   may be augmented (in the case of mrail).  Set during the
-	   transport's initialization, and should be read-only from
-	   that point. */
+	/*
+	 * name of the device - should include the provider name, but may be
+	 * augmented (in the case of mrail).  Set during the transport's
+	 * initialization, and should be read-only from that point.
+	 */
 	char *name;
 
+	/*
+	 * Protocol-agnostic MR cache for this device. Note that Registrations
+	 * are tied to domains in libfabric, but we do not have a
+	 * domain-specific object today, so stashing it in the device itself.
+	 * This should change if we were to break up nccl_net_ofi_device into
+	 * separate device and domain objects.
+	 */
+	nccl_ofi_mr_cache_t mr_cache;
```

Comment:
this more than doubles the size of the device and makes it span a second line (and the cache is splitting lines).

Some ideas on reducing this:

1. we can remove dev_id and it can be determined by scanning the plugin devices array and just comparing pointers. Can also pass it on the stack from plugin top-level call sites into each function, and/or rely on it already existing on the comm. Not strictly necessary, this is just 4 bytes.)
2. the plugin pointer is a global, I don't think we need for it to be in every device. I see no reason that should be faster.
3. the name pointers can also be made SoA and kept in the plugin itself beside the devices.
4. 24 bytes of statically-knowable function pointers at the tail here are the real killer.

Making those changes (move name to plugin, remove plugin pointer, resolve function pointers statically) can bring this back under a line. Alternatively, we can just move the cache to the end of the struct so that it's not splitting the line.

Reply:
I think this should be a pointer to the MR cache.  The MR cache isn't really critical path for most plugin send/recv (which are driven by collective ops and not nccl send/recv).  The critical path shouldn't involve dereferencing this pointer, so get it out of the critical memory path and move on.